File: bleep-core/src/scala/bleep/commands/LinkOptions.scala
package bleep
package commands

/** Options for controlling link behavior.
  *
  * These options control how Scala.js, Scala Native, Kotlin/JS, and Kotlin/Native projects are linked. Options that don't apply to a particular platform are
  * ignored.
  */
case class LinkOptions(
    /** Release mode - enables optimizations, disables debug info */
    releaseMode: Boolean,
    /** Generate source maps (JS platforms) */
    sourceMaps: Option[Boolean],
    /** Minify output (Scala.js only) */
    minify: Option[Boolean],
    /** Module kind for JS output */
    moduleKind: Option[LinkOptions.ModuleKind],
    /** Link-time optimization level (Scala Native only) */
    lto: Option[LinkOptions.LTO],
    /** Enable optimizations/DCE (all non-JVM platforms) */
    optimize: Option[Boolean],
    /** Include debug info (native platforms only) */
    debugInfo: Option[Boolean]
) {

  /** Convert to argument list for BSP protocol */
  def toArgs: List[String] = {
    val args = List.newBuilder[String]
    args += "--link"
    if (releaseMode) args += "--release"
    sourceMaps.foreach { v =>
      args += (if (v) "--source-maps" else "--no-source-maps")
    }
    minify.foreach { v =>
      args += (if (v) "--minify" else "--no-minify")
    }
    moduleKind.foreach { mk =>
      args += s"--module-kind=${mk.name}"
    }
    lto.foreach { l =>
      args += s"--lto=${l.name}"
    }
    optimize.foreach { v =>
      args += (if (v) "--optimize" else "--no-optimize")
    }
    debugInfo.foreach { v =>
      args += (if (v) "--debug-info" else "--no-debug-info")
    }
    args.result()
  }
}

object LinkOptions {

  /** Module kind for JavaScript output */
  sealed trait ModuleKind {
    def name: String
  }
  object ModuleKind {
    case object NoModule extends ModuleKind { val name = "nomodule" }
    case object CommonJS extends ModuleKind { val name = "commonjs" }
    case object ESModule extends ModuleKind { val name = "esmodule" }

    def fromString(s: String): Option[ModuleKind] = s.toLowerCase match {
      case "nomodule" => Some(NoModule)
      case "commonjs" => Some(CommonJS)
      case "esmodule" => Some(ESModule)
      case _          => None
    }

    val all: List[ModuleKind] = List(NoModule, CommonJS, ESModule)
  }

  /** Link-time optimization level (Scala Native) */
  sealed trait LTO {
    def name: String
  }
  object LTO {
    case object None extends LTO { val name = "none" }
    case object Thin extends LTO { val name = "thin" }
    case object Full extends LTO { val name = "full" }

    def fromString(s: String): Option[LTO] = s.toLowerCase match {
      case "none" => Some(None)
      case "thin" => Some(Thin)
      case "full" => Some(Full)
      case _      => scala.None
    }

    val all: List[LTO] = List(None, Thin, Full)
  }

  /** Default options (debug mode) */
  val Debug: LinkOptions = LinkOptions(
    releaseMode = false,
    sourceMaps = scala.None,
    minify = scala.None,
    moduleKind = scala.None,
    lto = scala.None,
    optimize = scala.None,
    debugInfo = scala.None
  )

  /** Release mode options */
  val Release: LinkOptions = LinkOptions(
    releaseMode = true,
    sourceMaps = scala.None,
    minify = scala.None,
    moduleKind = scala.None,
    lto = scala.None,
    optimize = scala.None,
    debugInfo = scala.None
  )
}

File: bleepscript/src/main/java/bleepscript/LinkOptions.java
package bleepscript;

import java.util.Objects;
import java.util.Optional;

/**
 * What {@link Commands#link} passes to the linker — the same settings {@code bleep link} accepts as
 * flags.
 *
 * <p>Every field a platform does not have a use for is ignored by it: {@code moduleKind} and {@code
 * minify} are JavaScript's, {@code lto} and {@code debugInfo} are the native platforms'. An empty
 * {@link Optional} means "whatever the build and the mode decide", which is not the same as an
 * explicit {@code false}: a project declaring {@code jsKind: esmodule} keeps its declaration when
 * {@code moduleKind} is empty and loses it when it is set.
 *
 * <p>{@link #DEBUG} and {@link #RELEASE} are the two most callers want.
 */
public record LinkOptions(
    boolean releaseMode,
    Optional<Boolean> sourceMaps,
    Optional<Boolean> minify,
    Optional<ModuleKind> moduleKind,
    Optional<LTO> lto,
    Optional<Boolean> optimize,
    Optional<Boolean> debugInfo) {

  public LinkOptions {
    Objects.requireNonNull(sourceMaps, "sourceMaps");
    Objects.requireNonNull(minify, "minify");
    Objects.requireNonNull(moduleKind, "moduleKind");
    Objects.requireNonNull(lto, "lto");
    Objects.requireNonNull(optimize, "optimize");
    Objects.requireNonNull(debugInfo, "debugInfo");
  }

  /** Module kind for JavaScript output (Scala.js). */
  public enum ModuleKind {
    NO_MODULE,
    COMMON_JS,
    ES_MODULE
  }

  /** Link-time optimization level (Scala Native). */
  public enum LTO {
    NONE,
    THIN,
    FULL
  }

  /** Debug semantics, nothing overridden. What {@code bleep link} does with no flags. */
  public static final LinkOptions DEBUG =
      new LinkOptions(
          false,
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Optional.empty());

  /** Release semantics, nothing overridden. What {@code bleep link --release} does. */
  public static final LinkOptions RELEASE =
      new LinkOptions(
          true,
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Optional.empty(),
          Optional.empty());
}
